src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -320,47 +326,63 @@ public WebSession authenticate(SessionManagementMethod sessionManagementMethod,
 		}
 		
 		/**
-		 * Modifies the input {@code requestMessage} by replacing fresh anti-CSRF token value in the request body.  
-		 * It first checks if the input {@code requestMessage} has an anti-CSRF token.
-		 * If yes, then it makes a {@code GET} request to the login page in order to renew the anti-CSRF token.
-		 * Then it modifies the input {@code requestMessage} with the regenerated anti-CSRF token value.
-		 * If the first {@code POST} login request does not have anti-CSRF token
+		 * <strong>Modifies</strong> the input {@code requestMessage} by replacing 
+		 * old ACSRF token value with the fresh one in the request body.  
+		 * It first checks if the input {@code loginMsgWithFreshAcsrfToken} has any anti-CSRF token.
+		 * If yes, then it modifies the input {@code requestMessage} with the fresh anti-CSRF token value.
+		 * If the {@code loginMsgWithFreshAcsrfToken} does not have any ACSRF token 
 		 * then the input {@code requestMessage} is left as it is.
 		 * 
-		 * @param requestMessage the login request message with correct credential
-		 * @param user the user of the context
+		 * @param requestMessage the login ({@code POST})request message with correct credentials
+		 * @param loginMsgWithFreshAcsrfToken the {@code HttpMessage} of the login page(/form) with fresh cookie and ACSRF token. 
 		 * @throws IOException if problem occurs when sending login request to get login page.
 		 */
-		private void replaceAntiCsrfTokenValueIfRequired(HttpMessage requestMessage, User user) throws IOException {
+		private void replaceAntiCsrfTokenValueIfRequired(HttpMessage requestMessage, HttpMessage loginMsgWithFreshAcsrfToken) throws IOException {
 			if(extAntiCsrf == null) {
 				extAntiCsrf = Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.class);
 			}
-			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+			List<AntiCsrfToken> freshAcsrfTokens = null;
   			if(extAntiCsrf != null) {
-				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
-			}
-			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
-			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
-				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+				freshAcsrfTokens = extAntiCsrf.getTokensFromResponse(loginMsgWithFreshAcsrfToken);
+			} else {
+				LOGGER.warn("ExtensionAntiCSRF is not available, skipping ACSRF replacing task");
+				return;
 			}
-			
-			if (antiCsrfTokenOfFirstLoginMsg != null) {
-				HttpMessage loginMsgToRegenerateAntiCsrfToken = antiCsrfTokenOfFirstLoginMsg.getMsg().cloneAll();
-				loginMsgToRegenerateAntiCsrfToken.setRequestingUser(user);
-				
-				// Clear any session identifiers
-				loginMsgToRegenerateAntiCsrfToken.getRequestHeader().setHeader(HttpRequestHeader.COOKIE, null);
+			if(freshAcsrfTokens != null && freshAcsrfTokens.size() > 0) {
```

Comment:
We could invert the condition and return sooner (like done in the previous if), to reduce nesting.